src/api/initialDataTypes.js

```diff
@@ -499,13 +498,47 @@ export type InitialDataUpdateMessageFlags = $ReadOnly<{|
   |}>,
 |}>;
 
+/**
+ * Current values for (all?) users' self-assigned availability and
+ *   text/emoji statuses
+ */
+// TODO: Find out if servers ignore any users (e.g., deactivated ones) when
+//   assembling this.
 export type InitialDataUserStatus = $ReadOnly<{|
   /**
    * Older servers (through at least 1.9.1) don't send this.
-   * A missing value is equivalent to empty.
```

Comment:
types [nfc]: Remove misleading bit of jsdoc on initial-data `user_status`
    
    Looks like this was added in c8d071452.
    
    I think it's unlikely that `user_status` has ever been missing
    except because the app is talking to an old server that doesn't
    support `user_status`.

Yeah, I believe what was meant by this line was: if this is missing, that means this is an old server which doesn't have the `user_status` feature, and then the meaning of that is the same as if we got a `user_status` value that was empty.


    A missing `user_status` from an old server doesn't mean the same
    thing as a present but empty `user_status` from a current server.
    
    In particular:
    
    - a missing `user_status` from an old server means we just don't
      know what users' statuses are.

Hmm. I don't think I quite agree with this interpretation.

If the server doesn't have the feature of user status, then I would say that there is no such thing as the status of a user on that server. It's not that they really do have a status of ":calendar: In a meeting" or ":palm_tree: Vacationing" or whatever and we don't know what it is -- we do know the status of each user, and its value is… no-status.

It's true that out in the real world, the human behind that user may really be in a meeting, or vacationing, or so on, and we don't know which of those they are. But it's not like we know that real-world state, or even have all that solid a guess at it, when the feature does exist. (People tend to set these things and not update them. Or interpreted another way, they use them abstractly as signifiers -- they leave it at "In a meeting", say, not because they're literally always in meetings, nor by carelessness, but because they want to discourage you from PMing them and expecting a quick response.) So the "user status" we're concerned with isn't that real-world state of the human in any case -- rather it's whatever value they've chosen to communicate through this product feature. Which, when the feature doesn't exist, is no value.

Now the question is, how should we handle having no status value?

    - a present but empty `user_status` from a current server means
      everyone's status is boring (in particular, everyone is
      "available" and no one currently has a text or emoji status set);
      see
        https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/Emoji.20statuses.20in.20zulip.2Eyaml/near/1322329 .

I don't think it's right to say that a boring status, on a modern server where user status exists, means the person is "available". Here's the user-facing doc:
https://zulip.com/help/status-and-availability

> Your **availability** is a colored dot (like :green_circle:) that indicates if you're currently active on Zulip, idle, offline, or unavailable.

Then it goes on to define the conditions for each of those "four possible availabilities". We find that a user's "availability" is defined by a combination of presence with whether they've manually "set [themself] as unavailable". The latter refers to the `away` property in `user_status`.

So with a boring status -- `away` false, represented in initial data by being absent -- their availability is determined purely by presence. That's exactly what we want to do in the absence of the user_status feature, too.

The same thing ends up holding for the other components of user status, namely the text and the emoji: when they haven't been set by the user, the thing we want to do is exactly the same as if the user status feature didn't exist.

That's what this comment was saying, though too tersely: the semantics of the user-status feature is such that when it's unset for a given user, the various things it would affect behave as if the feature didn't exist at all.

Reply:
> * Is "boring" a helpful label for anything here, or should we choose something else? In this draft I've got a `kBoringUserStatus`. Are you thinking that should change?

In a different API discussion recently I called something similar a "zero" value:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/stream.20IDs.20in.20.60muted_topics.60/near/1325745
By that I mean the value that something has when there isn't any explicit record saying what value it has.

In that other context it's fundamental that there has to be a zero value, because that was about a relation between users and topics, and if we tried to explicitly store a value for every pair of a user and a topic that would be an enormous, quadratic amount of space. So we can't do that, so there must be many (user, topic) pairs with no explicit value, so there must be some interpretation of what that means, and that interpretation is the zero value.

Here, there didn't absolutely *have* to be a zero value -- it'd be perfectly feasible for every user to have an explicit status. But the design we chose does have a zero value. The API docs aren't real clear on this point, hmm -- particularly at `user_status` in https://zulip.com/api/register-queue -- but that does say:
> A dictionary which contains the status of all users in the Zulip organization who have set a status.

which means there can be users who have *not* "set a status", there won't be an explicit record for them, and there must be some interpretation of what that means. (And in [this chat message](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/Emoji.20statuses.20in.20zulip.2Eyaml/near/1322329) you laid out what the facts seem to be of that interpretation, which the docs should be amended to include.) So whatever the status is for users who have not "set a status" and don't appear in that data -- that's the zero status.

And that zero value is a bit more than just an arbitrary optimization in the data, too -- it reflects something real in the UX of this product feature. The zero user-status is the same as the user-status you have if you've just created your account and never interacted with the feature. I.e. it's the default, but moreover this isn't like language or something where there are many possible defaults which might vary between servers or realms -- there's really one canonical default that makes sense given the nature of this product feature. (I guess *maybe* there could be an org that wants a default `away` of true; but to have every user start with some status emoji, or some non-empty status text, would really not make sense unless it were an attempt to redefine the feature to serve some purpose it wasn't intended for.) That fact gives it a bit more weight and permanence in the API than if it were just an arbitrary optimization… though really even without that, so long as it is specified there in the API, that's enough already to make it legit for us to rely on it.

Anyway, if that discussion makes sense to you, then perhaps "zero" is a good term for more precisely getting at basically the same thing "boring" is expressing here, and it could be `kZeroUserStatus`.

Reply:
> However, I think it's possible that data is omitted in the first place (by status-aware servers) just because servers want to minimize how much data they send while still giving complete information.

Yeah, that is possible in principle, and there could be some feature in the future where there's a zero value in how the data's expressed (by servers that have the feature) and some different value that we want to use for handling a server that doesn't have the feature.

Here, though, those two values coincide. And while it doesn't have to *always* be that way, I think it's far from an accident that they do:
* When there's a zero value (in the sense of sometimes having no explicit record, and a value to interpret that as implicitly meaning), it's generally going to correspond to what happens when a user hasn't chosen to actively interact with something. After all, that's generally going to be the common case.
* Whatever happens when a user hasn't chosen to actively interact with something via some feature, that's generally going to be the same thing that happened with servers that didn't have the feature at all. That is, new features are much more often opt-in than opt-out.

In any case, it is true of *this* feature that the value to use in those two situations is the same; so that's something we can rely on.

And then it's not particularly odd that that's the case -- it's the way things can typically be expected to be. So it's good to have a brief comment noting it explicitly (to confirm that we did think about this question and check that this agreement holds for this feature), but then we can proceed to cheerfully and matter-of-factly rely on it.

Reply:
> We do have to choose whether or not to offer a UI to set these status components. We shouldn't show an emoji picker if the server doesn't support emoji statuses, and we shouldn't try to let people set themselves "unavailable" or choose a text status when those features aren't supported.
> 
> Happily, we have the feature level, so we don't have to go back to the edge to check for presence/shape of the status data we got straight from the server. slightly_smiling_face

Yep, agreed.